src/partest-extras/scala/tools/partest/async/AsyncStateMachine.scala

```diff
@@ -0,0 +1,35 @@
+/*
+ * Scala (https://www.scala-lang.org)
+ *
+ * Copyright EPFL and Lightbend, Inc.
+ *
+ * Licensed under Apache License 2.0
+ * (http://www.apache.org/licenses/LICENSE-2.0).
+ *
+ * See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.
+ */
+
+package scala.tools.partest.async
+
+// The async phase expects the state machine class to structurally conform to this interface.
+trait AsyncStateMachine[F, R] {
+  /** Assign `i` to the state variable */
+  protected def state$async_=(i: Int): Unit
+  /** Retrieve the current value of the state variable */
+  protected def state$async: Int
```

Comment:
Could we not use `$` in method names that users will have to see at some point? `$` are supposed to be reserved for the compiler, so we shouldn't encourage people by showing them us using such `$`s.